packages/flutter/lib/src/material/text_selection.dart

```diff
@@ -54,9 +54,9 @@ class MaterialTextSelectionControls extends TextSelectionControls {
 
   /// Builder for material-style text selection handles.
   @override
-  Widget buildHandle(BuildContext context, TextSelectionHandleType type, double textHeight, [VoidCallback? onTap, double? startGlyphHeight, double? endGlyphHeight]) {
-    final ThemeData theme = Theme.of(context);
-    final Color handleColor = TextSelectionTheme.of(context).selectionHandleColor ?? theme.colorScheme.primary;
+  Widget buildHandle(BuildContext textContext, TextSelectionHandleType type, double textHeight, [VoidCallback? onTap, double? startGlyphHeight, double? endGlyphHeight]) {
+    final ThemeData theme = Theme.of(textContext);
```

Comment:
@chunhtai brought up that if the selection overlay is removed in `EditableTextState.dispose`, the overlay entry will still be in the widget tree and if `buildHandle` gets called at that point `Theme.of` here and the `TextSelectionTheme.of` will throw because `textContext` is no longer mounted.

Reply:
https://github.com/flutter/flutter/blob/ca14d85cf7f983ce038f0f385283e314ea55afdf/packages/flutter/lib/src/widgets/overlay.dart#L154

The entry will be removed from the overlay state in this frame but it would rebuild until next, so the overlay entry will out live editable text for one frame. It will have to be a really niche case this can go wrong. For example, the overlay entry some how rebuild it self within this frame. Is it possible to pass in the theme instead? or alternatively builds the theme right on top of the selection overlay.

Reply:
If `_handleSelectionChanged` and `dispose` is called at the same frame (which seems to be possible when the text field receives a method channel call during the idle phase (and `_handleSelectionChanged` will be called in response to that), and the text field then proceeds to get unmounted in the same frame) then this can throw? 

You might be able to reproduce the problem by creating a widget test that does the following:
1.  pump a widget tree with a focused text field.
2. send the text field a method call (or just call `EditableTextState.updateEditingValue` directly) to update the selection to schedule a rebuild for the selection handles
3. pump the same widget tree without that text field.

Reply:
I tried to reproduce this in various ways but I just don't seem be able to. Do you see anything wrong with this?
```dart
  testWidgets('TextSelectionTheme used for handles with disposed context', (WidgetTester tester) async {
    EditableText.debugDeterministicCursor = true;
    addTearDown(() {
      EditableText.debugDeterministicCursor = false;
    });

    // Test TextField's cursor & selection color.
    final TextEditingController editingController = TextEditingController(text: 'abc def ghi');
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: TextField(
            controller: editingController,
            autofocus: true,
          ),
        ),
      ),
    );

    // Long press the 'e' to focus text field and select 'def'.
    const int tapIndex = 5;
    final Offset ePos = textOffsetToPosition(tester, tapIndex);
    await tester.longPressAt(ePos);

    // Wait for controls to appear
    await tester.pumpAndSettle();

    // Update editing value to schedule a rebuild of the selection handles
    final EditableTextState selection = tester.state(find.byType(EditableText));
    selection.updateEditingValue(selection.currentTextEditingValue.copyWith(composing: const TextRange(start: 0, end: 10)));

    // Pump the same widget tree without the text field
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: Container(),
        ),
      ),
    );
  });
```

FYI since this is used by both material and cupertino (In noticed it fixes the faulty behavior for both by the way) then it isn't really possible to import either library in `widgets/text_selection.dart` to pass the theme or build it around the `_TextSelectionHandleOverlay`. One way to do this would be to update a theme field in `MaterialTextSelectionControls` and `CupertinoTextSelectionControls`? Doesn't seem ideal and I don't exactly know how to check if the context is valid for updating this field.

Reply:
`selection.updateEditingValue(selection.currentTextEditingValue.copyWith(composing: const TextRange(start: 0, end: 10)));` it seems you're only changing the composing region?

Reply:
This does schedule a rebuild though

Reply:
Using the sibling tree seems pretty dangerous to me even if you can't reproduce it now.

I can't think of a way to refactor this unless we have a common theme interface in widget layer....

Reply:
I agree that it seems dangerous. Aside from `InheritedTheme` I don't think there is anything in the widget layer?

Reply:
Yeah it looks like `InheritedTheme` is a safer alternative.